src/tactic/core.lean

```diff
@@ -1496,41 +1481,73 @@ The number 2 indicates that first the two bound variables of the
 rather than `=ₐ` or `is_def_eq` tells us that proof script should
 not see the difference between the two.
  -/
-meta def goal := ℕ × expr
+meta def user_visible_goal := ℕ × expr
 
 /-- proof state made of multiple `goal` meant for comparing
 the result of running different tactics -/
-meta def proof_state := list goal
+meta def proof_state := list user_visible_goal
 
-meta instance goal.inhabited : inhabited goal := ⟨(0,var 0)⟩
+meta instance goal.inhabited : inhabited user_visible_goal := ⟨(0,var 0)⟩
 meta instance proof_state.inhabited : inhabited proof_state :=
-(infer_instance : inhabited (list goal))
+(infer_instance : inhabited (list user_visible_goal))
 
-/--
-Run `tac` in a disposable proof state and return the state.
-See `proof_state` and `goal`.
+meta def get_user_visible_goal : tactic user_visible_goal :=
+retrieve $ do
+ls ← local_context,
+     tgt ← target >>= instantiate_mvars >>= pis ls,
+     pure (ls.length, tgt)
+
+meta def goal_of_mvar (g : expr) : tactic user_visible_goal :=
+with_local_goals' [g] get_user_visible_goal
+
+/-- `get_proof_state` lists the user visible goal for each goal
+    of the current state and for each goal, abstracts all of the
+    meta variables of the other gaols.
+
+    This produces a list of goals in the form of `ℕ × expr` where
+    the `expr` encodes the following proof state:
+
+      ```lean
+      2 goals
+      l₁ : t₁,
+      l₂ : t₂,
+      l₃ : t₃
+      ⊢ tgt₁
 
-Limitation:
-If one goal occurs as a meta variable in another goal, the representation
-that this function returns will not compare goals have in the desired way.
+      ⊢ tgt₂
+      ```
 
-Possible improvement:
-If one wants to account for that situation, `kabstract` can be used to replace
-the meta variables that are goals in the final state with bound variables.
-Then, if `(3, p)` is a goal in `gs`, `p` has `3 + gs.length` bound variables
-that do not stand for universal quantifications in the goal.
+      as
+
+      ```lean
+      [ (3, ∀ (mv : tgt₁) (mv : tgt₂) (l₁ : t₁) (l₂ : t₂) (l₃ : t₃), tgt₁),
+        (0, ∀ (mv : tgt₁) (mv : tgt₂), tgt₂) ]
+      ```
+
+      with 2 goals, the first 2 bound variables encode the meta variable
+      of all the goals, the next 3 (in the first goal) and 0 (in the second goal)
+      are the local constants.
+
+      This representation allows us to compare goals and proof states while
+      ignoring information like the unique name of local constants and
+      the equality or difference of meta variables that encode the same goal.
+ -/
+meta def get_proof_state : tactic proof_state :=
+do gs ← get_goals,
+   gs.mmap $ λ g, do
+     ⟨n,g⟩ ← goal_of_mvar g,
+     g ← gs.mfoldl (λ g v, do
```

Comment:
Wait, isn't this now worse than before?  `tgt₁` and `tgt₂` will typically contain local constants (e.g. from `cases`, and they will change on every tactic invocation).

I wouldn't spend too much effort on properly supporting metavariables.

Reply:
Yes you're right because the type of the bound variables will refer to transient information. We can use `kabstract` and not bind the resulting variables. Or we can bind them and give them `true` as type.

Reply:
How important is this actually for `squeeze_simp`?  Normalizing the metavariables is only important in the rare case when the metavariables are introduced inside the `squeeze_simp`, right?  Like `squeeze_simp { existsi _, simp, apply default }`?

> Or we can bind them and give them true as type.

Yes, that should actually work.  The resulting term won't be well-typed, but we don't care.

We could also replace them by local constants (i.e. `_goal.0`, `_goal.1`, etc.), then we avoid the (IMHO ugly) binders.